`operator||` causes a compilation error when compiling with `-fpreview-breaking-changes`

**Describe the bug**
Adding the `-fpreview-breaking-changes` flag causes a compilation error when using boolean logic operators defined in the `sycl::vec` class. Compiling without the flag doesn't produce any errors.

**To Reproduce**
```c++
#include <sycl/sycl.hpp>

int main() {
  sycl::queue q;
  q.single_task([=]() {
    auto testVec1 = sycl::vec<double, 16>(static_cast<double>(1));
    auto testVec2 = sycl::vec<double, 16>(static_cast<double>(2));
    sycl::vec<std::int64_t, 16> resVec = testVec1 || testVec2;
  });
}
```

```
$ clang++ -fsycl ./reproducer.cpp
$ clang++ -fsycl -fpreview-breaking-changes ./reproducer.cpp
In file included from ./reproducer.cpp:1:
In file included from /llvm/build/bin/../include/sycl/sycl.hpp:11:
In file included from /llvm/build/bin/../include/sycl/accessor.hpp:15:
In file included from /llvm/build/bin/../include/sycl/buffer.hpp:13:
In file included from /llvm/build/bin/../include/sycl/context.hpp:16:
In file included from /llvm/build/bin/../include/sycl/detail/info_desc_helpers.hpp:20:
In file included from /llvm/build/bin/../include/sycl/info/info_desc.hpp:20:
In file included from /llvm/build/bin/../include/sycl/ext/oneapi/matrix/query-types.hpp:11:
In file included from /llvm/build/bin/../include/sycl/ext/oneapi/bfloat16.hpp:12:
In file included from /llvm/build/bin/../include/sycl/builtins.hpp:14:
In file included from /llvm/build/bin/../include/sycl/builtins_marray_gen.hpp:15:
In file included from /llvm/build/bin/../include/sycl/builtins_vector_gen.hpp:14:
In file included from /llvm/build/bin/../include/sycl/builtins_utils_vec.hpp:14:
/llvm/build/bin/../include/sycl/types.hpp:1119:3: error: value of type 'const DataType' (aka 'const std::array<double, 16>') is not contextually convertible to 'bool'
 1119 |   __SYCL_RELLOGOP(||)
      |   ^~~~~~~~~~~~~~~~~~~
/llvm/build/bin/../include/sycl/types.hpp:1078:13: note: expanded from macro '__SYCL_RELLOGOP'
 1078 |             m_Data RELLOGOP Rhs.m_Data));                                      \
      |             ^~~~~~
./reproducer.cpp:8:51: note: in instantiation of member function 'sycl::vec<double, 16>::operator||' requested here
    8 |     sycl::vec<std::int64_t, 16> resVec = testVec1 || testVec2;
      |                                                   ^
In file included from ./reproducer.cpp:1:
In file included from /llvm/build/bin/../include/sycl/sycl.hpp:11:
In file included from /llvm/build/bin/../include/sycl/accessor.hpp:15:
In file included from /llvm/build/bin/../include/sycl/buffer.hpp:13:
In file included from /llvm/build/bin/../include/sycl/context.hpp:16:
In file included from /llvm/build/bin/../include/sycl/detail/info_desc_helpers.hpp:20:
In file included from /llvm/build/bin/../include/sycl/info/info_desc.hpp:20:
In file included from /llvm/build/bin/../include/sycl/ext/oneapi/matrix/query-types.hpp:11:
In file included from /llvm/build/bin/../include/sycl/ext/oneapi/bfloat16.hpp:12:
In file included from /llvm/build/bin/../include/sycl/builtins.hpp:14:
In file included from /llvm/build/bin/../include/sycl/builtins_marray_gen.hpp:15:
In file included from /llvm/build/bin/../include/sycl/builtins_vector_gen.hpp:14:
In file included from /llvm/build/bin/../include/sycl/builtins_utils_vec.hpp:14:
/llvm/build/bin/../include/sycl/types.hpp:1119:19: error: invalid operands to binary expression ('const DataType' (aka 'const std::array<double, 16>') and 'const DataType')
 1119 |   __SYCL_RELLOGOP(||)
      |   ~~~~~~~~~~~~~~~~^~~
/llvm/build/bin/../include/sycl/types.hpp:1078:20: note: expanded from macro '__SYCL_RELLOGOP'
 1078 |             m_Data RELLOGOP Rhs.m_Data));                                      \
      |             ~~~~~~ ^        ~~~~~~~~~~
2 errors generated.
```

**Environment (please complete the following information):**

- OS: Linux
- Target device and vendor: N/A
- DPC++ version: https://github.com/intel/llvm/commit/ff61613e225a77913f862f0849ed8a36d7dc14aa
- Dependencies version: N/A

**Additional context**
I think this error is related to changes in https://github.com/intel/llvm/pull/11697. The `sycl::vec` class is using an `std::array` internally, but the latter doesn't implement `operator||`.

--- a/sycl/include/sycl/types.hpp
+++ b/sycl/include/sycl/types.hpp
@@ -1074,7 +1074,7 @@ public:
 // by SYCL device compiler only.
 #ifdef __SYCL_DEVICE_ONLY__
 #define __SYCL_RELLOGOP(RELLOGOP)                                              \
-  vec<rel_t, NumElements> operator RELLOGOP(const vec &Rhs) const {            \
+  vec<rel_t, NumElements> operator RELLOGOP(const vec & Rhs) const {           \
     vec<rel_t, NumElements> Ret{};                                             \
     /* This special case is needed since there are no standard operator||   */ \
     /* or operator&& functions for std::array.                              */ \
@@ -1100,7 +1100,7 @@ public:
                                 (std::is_fundamental_v<vec_data_t<T>> ||       \
                                  std::is_same_v<T, half>),                     \
                             vec<rel_t, NumElements>>                           \
-  operator RELLOGOP(const T &Rhs) const {                                      \
+  operator RELLOGOP(const T & Rhs) const {                                     \
     return *this RELLOGOP vec(static_cast<const DataT &>(Rhs));                \
   }
 #else

--- a/sycl/include/sycl/types.hpp
+++ b/sycl/include/sycl/types.hpp
@@ -1074,12 +1074,25 @@ public:
 // by SYCL device compiler only.
 #ifdef __SYCL_DEVICE_ONLY__
 #define __SYCL_RELLOGOP(RELLOGOP)                                              \
-  vec<rel_t, NumElements> operator RELLOGOP(const vec & Rhs) const {           \
-    auto Ret =                                                                 \
-        vec<rel_t, NumElements>((typename vec<rel_t, NumElements>::vector_t)(  \
-            m_Data RELLOGOP Rhs.m_Data));                                      \
-    if (NumElements == 1) /*Scalar 0/1 logic was applied, invert*/             \
-      Ret *= -1;                                                               \
+  vec<rel_t, NumElements> operator RELLOGOP(const vec &Rhs) const {            \
+    vec<rel_t, NumElements> Ret{};                                             \
+    /* This special case is needed since there are no standard operator||   */ \
+    /* or operator&& functions for std::array.                              */ \
+    if constexpr (IsUsingArrayOnDevice &&                                      \
+                  (std::string_view(#RELLOGOP) == "||" ||                      \
+                   std::string_view(#RELLOGOP) == "&&")) {                     \
+      for (size_t I = 0; I < NumElements; ++I) {                               \
+        Ret.setValue(I,                                                        \
+                     -(vec_data<DataT>::get(getValue(I))                       \
+                           RELLOGOP vec_data<DataT>::get(Rhs.getValue(I))));   \
+      }                                                                        \
+    } else {                                                                   \
+      Ret = vec<rel_t, NumElements>(                                           \
+          (typename vec<rel_t, NumElements>::vector_t)(                        \
+              m_Data RELLOGOP Rhs.m_Data));                                    \
+      if (NumElements == 1) /*Scalar 0/1 logic was applied, invert*/           \
+        Ret *= -1;                                                             \
+    }                                                                          \
     return Ret;                                                                \
   }                                                                            \
   template <typename T>                                                        \
@@ -1087,7 +1100,7 @@ public:
                                 (std::is_fundamental_v<vec_data_t<T>> ||       \
                                  std::is_same_v<T, half>),                     \
                             vec<rel_t, NumElements>>                           \
-  operator RELLOGOP(const T & Rhs) const {                                     \
+  operator RELLOGOP(const T &Rhs) const {                                      \
     return *this RELLOGOP vec(static_cast<const DataT &>(Rhs));                \
   }
 #else

--- a/sycl/test-e2e/Regression/vec_logical_ops.cpp
+++ b/sycl/test-e2e/Regression/vec_logical_ops.cpp
@@ -1,0 +1,15 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+// RUN: %if preview-breaking-changes-supported %{ %clangxx -fsycl -fpreview-breaking-changes && ./%t.out %}
+
+#include <sycl/sycl.hpp>
+
+int main() {
+  sycl::queue q;
+  q.single_task([=]() {
+    auto testVec1 = sycl::vec<double, 16>(static_cast<double>(1));
+    auto testVec2 = sycl::vec<double, 16>(static_cast<double>(2));
+    sycl::vec<std::int64_t, 16> resVec1 = testVec1 || testVec2;
+    sycl::vec<std::int64_t, 16> resVec2 = testVec1 && testVec2;
+  });
+}

